qa/task-cancellation/src/test/java/org/elasticsearch/tasks/TaskCancellationIT.java

```diff
@@ -0,0 +1,180 @@
+            }
+        });
+    }
+
+    Set<String> getNodeIds(RestHighLevelClient client) throws IOException {
+        Response response = client.getLowLevelClient().performRequest(new Request("GET", "_nodes"));
+        ObjectPath objectPath = ObjectPath.createFromResponse(response);
+        final Map<String, Object> nodeMap = objectPath.evaluate("nodes");
+        return nodeMap.keySet();
+    }
+
+    @Before
+    public void initClientsAndClusters() throws Exception {
+        RestClientBuilder.RequestConfigCallback callback = config -> config.setSocketTimeout(120 * 1000);
+        oldCluster = new RestHighLevelClient(
+            RestClient.builder(readHosts("tests.rest.old_cluster").toArray(new HttpHost[0])).setRequestConfigCallback(callback)
+        );
+        newCluster = new RestHighLevelClient(
+            RestClient.builder(readHosts("tests.rest.cluster").toArray(new HttpHost[0])).setRequestConfigCallback(callback)
+        );
+        // connect the new cluster to the old one
+        configureRemoteCluster(newCluster, "old", randomSubsetOf(1, readHosts("tests.old_cluster")));
+    }
+
+    @After
+    public void destroyClients() throws IOException {
+        IOUtils.close(oldCluster, newCluster);
+    }
+
+    public void testCancelTasks() throws Exception {
+        Collection<String> newNodes = randomSubsetOf(getNodeIds(newCluster));
+        Collection<String> oldNodes = randomSubsetOf(between(1, 2), getNodeIds(oldCluster));
+        Thread thread = new Thread(() -> {
+            final ResponseException exception = expectThrows(ResponseException.class, () -> {
+                final Request request = new Request("POST", "/_test_blocking");
+                request.addParameter("id", "1");
+                request.addParameter(
+                    "targets",
+                    Stream.concat(newNodes.stream().map(n -> ":" + n), oldNodes.stream().map(n -> "old:" + n))
+                        .collect(Collectors.joining(","))
+                );
+                newCluster.getLowLevelClient().performRequest(request);
+            });
+            try {
+                final String reason = EntityUtils.toString(exception.getResponse().getEntity());
+                assertThat(reason, containsString("\"type\":\"task_cancelled_exception\""));
+            } catch (IOException ex) {
+                throw new UncheckedIOException(ex);
+            }
+        });
+        thread.start();
+        try {
+            assertBusy(() -> {
+                List<TaskInfo> oldTasks = oldCluster.tasks()
+                    .list(new ListTasksRequest().setActions("internal::test_action"), RequestOptions.DEFAULT)
+                    .getTasks()
+                    .stream()
+                    .filter(t -> oldNodes.contains(t.getTaskId().getNodeId()))
+                    .collect(Collectors.toList());
+                assertThat(oldTasks.size(), greaterThanOrEqualTo(oldNodes.size()));
```

Comment:
why might there be more tasks than oldNodes? Is it because the  request might be rerouted through such an old node if it has remote role?

Reply:
This assertion is not strong enough. I pushed https://github.com/elastic/elasticsearch/pull/59828/commits/1cf007485ad37e12a0bf6aced95442126dab46cb.